File: python/draw_ctcr.py
import numpy as np
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d.art3d import Poly3DCollection

from utils import robotindependentmapping, setupfigure


def draw_ctcr(g: np.ndarray[np.ndarray[float]], tube_end: np.ndarray[int], r_tube: np.ndarray[float], tipframe: bool=True, segframe: bool=False, baseframe: bool=False, projections: bool=False, baseplate: bool=True) -> None:
    '''
    DRAW_CTCR Creates a figure of a concentric tube continuum robot (ctcr)

    Takes a matrix with nx16 entries, where n is the number
    of points on the backbone curve. For each point on the curve, the 4x4
    transformation matrix is stored columnwise (16 entries). The x- and
    y-axis span the material orientation and the z-axis is tangent to the
    curve.

    Parameters
    ----------
    g: ndarray
        Backbone curve with n 4x4 transformation matrices reshaped into 1x16 vector (columnwise)
    tube_end: ndarray
        Indices of g where ctcr tubes terminate
    r_tube: ndarray
        Radii of tubes
    tipframe: bool, default=True
        Shows tip frame
    segframe: bool, default=False
        Shows segment end frames
    baseframe: bool, default=False
        Shows robot base frame
    projections: bool, default=False
        Shows projections of backbone curve onto coordinate axes
    baseplate: bool, default=True
        Shows robot base plate

    Outputs
    -------
    A matplotlib figure object
        Figure object of the generated plot

    Author: Jessica Burgner-Kahrs <jbk@cs.toronto.edu>
    Date: 2023/02/16
    Version: 0.2
    Copyright: 2023 Continuum Robotics Laboratory, University of Toronto
    '''

    if np.max(tube_end) > g.shape[0] or tube_end.size != r_tube.size:
        raise ValueError("Dimension mismatch")

    # Setup figures
    ax = setupfigure(g=g, seg_end=tube_end, tipframe=tipframe, segframe=segframe, baseframe=baseframe, projections=projections, baseplate=baseplate)

    numtubes = tube_end.size
    radial_pts = 16  # resolution (num point on circular cross section, increase for high detail level)
    tcirc = np.linspace(0, 2*np.pi, radial_pts)
    alpha = 1  # 0 = transparent

    if numtubes == 1:
        col = np.array([0.8])
    else:
        col = np.linspace(0.2, 0.8, numtubes)

    ## draw tubes
    start_tube = 0
    for j in range(numtubes):
        end_tube = tube_end[j]
        color = [col[j], col[j], col[j]]

        # points on a circle in the local x-y plane
        basecirc = np.vstack([r_tube[j] * np.sin(tcirc), r_tube[j] * np.cos(tcirc), np.zeros(radial_pts), np.ones(radial_pts)])

        # Loop to draw each cylindrical segment for tube
        for i in range(start_tube, end_tube - 1):
            # Get the coordinates of the current segment
            T_point = g[i, :].reshape((4, 4)).T
            basecirc_trans = T_point @ basecirc

            # Get the coordinates of the next segment
            T_point = g[i + 1, :].reshape((4, 4)).T
            basecirc_trans_ahead = T_point @ basecirc

            # Loop to draw each radial point for this entire segment
            for k in range(radial_pts - 1):
                xedge = np.array([basecirc_trans[0, k], basecirc_trans[0, k + 1], basecirc_trans_ahead[0, k + 1], basecirc_trans_ahead[0, k]])
                yedge = np.array([basecirc_trans[1, k], basecirc_trans[1, k + 1], basecirc_trans_ahead[1, k + 1], basecirc_trans_ahead[1, k]])
                zedge = np.array([basecirc_trans[2, k], basecirc_trans[2, k + 1], basecirc_trans_ahead[2, k + 1], basecirc_trans_ahead[2, k]])

                verts = [list(zip(xedge, yedge, zedge))]
                ax.add_collection3d(Poly3DCollection(verts, alpha=alpha, color=color, rasterized=True, zorder=10), zdir='z')

        start_tube = end_tube

    plt.show()


if "__main__" == __name__:
    kappa = np.array([1/30e-3, 1/40e-3, 1/15e-3])
    phi = np.array([0, np.deg2rad(160), np.deg2rad(30)])
    ell = np.array([50e-3, 70e-3, 25e-3])
    pts_per_seg = np.array([30])

    g = robotindependentmapping(kappa, phi, ell, pts_per_seg)
    draw_ctcr(g, np.array([30, 60, 90]), np.array([2e-3, 1.5e-3, 1e-3]), tipframe=True, segframe=True, baseframe=True, projections=True, baseplate=True)

File: matlab/ctcr/draw_ctcr.m
function [fig] = draw_ctcr(g,tube_end,r_tube,options)
% DRAW_CTCR Creates a figure of a concentric tube continuum robot (ctcr)
%
%   Takes a matrix with nx16 entries, where n is the number
%   of points on the backbone curve. For each point on the curve, the 4x4
%   transformation matrix is stored columnwise (16 entries). The x- and
%   y-axis span the material orientation and the z-axis is tangent to the
%   curve.
%
%   INPUT
%   g(n,16): backbone curve with n 4x4 transformation matrices reshaped into 1x16 vector (columnwise)
%   tube_end(1,m): indices of g where ctcr tubes terminate
%   r_tube(1,m): radii of tubes
%   options:
%       tipframe (shows tip frame, default true/1)
%       baseframe (shows robot base frame, default false/0)
%       projections (shows projections of backbone curve onto
%                    coordinate axes, default false/0)
%       baseplate (shows robot base plate, default false/0)
%
%
%   Author: Jessica Burgner-Kahrs <jbk@cs.toronto.edu>
%   Date: 2023/02/16
%   Version: 0.2
%
%   Copyright: 2023 Continuum Robotics Laboratory, University of Toronto

    arguments
        g (:,16) double %backbone curve (transformation matrices stored columnwise)
        tube_end (1,:) uint8 %tube end indices
        r_tube (1,:) double % tube radius
        options.tipframe (1,1) {mustBeNumericOrLogical} = 1
        options.segframe (1,1) {mustBeNumericOrLogical} = 0
        options.baseframe (1,1) {mustBeNumericOrLogical} = 0
        options.projections (1,1) {mustBeNumericOrLogical} = 0
        options.baseplate (1,1) {mustBeNumericOrLogical} = 1
    end

    if max(tube_end)>size(g,1) || size(tube_end,2) ~= size(r_tube,2)
        error("Dimension mismatch")
    end

    curvelength = sum(vecnorm(g(2:end,13:15)'-g(1:end-1,13:15)'));
    numtubes = size(tube_end,2);

    %% Setup figure
    fig=figure;
    hold on
    set(fig,'Position',[0 0 1280 1024]);

    % Axes, Labels
    clearance = 0.03;
    axis([-(max(abs(g(:,13)))+clearance) max(abs(g(:,13)))+clearance ...
          -(max(abs(g(:,14)))+clearance) max(abs(g(:,14)))+clearance ...
          0 curvelength+clearance])
    ax=gca;
    ax.GridAlpha=0.3;

    xlabel('x (m)')
    ylabel('y (m)')
    zlabel('z (m)')
    grid on
    view([0.5 0.5 0.5])
    daspect([1 1 1])
    hold on

    radial_pts = 16; %resolution (num point on circular cross section, increase for high detail level
    tcirc = linspace(0,2*pi,radial_pts);
    col = linspace(0.2,0.8,numtubes);
    alpha = 1;%0 = transparent

    %% draw tubes
    for j=1:numtubes
        %set tube start and end index for current tube
        if j == 1
            starttube = 1;
        else
            starttube=tube_end(j-1);
        end
        endtube = tube_end(j);
        color = col(j)*[1,1,1];

        % points on a circle in the local x-y plane
        basecirc = [r_tube(j)*sin(tcirc);r_tube(j)*cos(tcirc);zeros(1,length(tcirc));ones(1,length(tcirc))];

        % transform circle points into the tube's base frame
        basecirc_trans = reshape(g(starttube,:),4,4)*basecirc;

        %draw patches to fill in the circle
        patch(basecirc_trans(1,:),basecirc_trans(2,:),basecirc_trans(3,:),color,'EdgeAlpha',0,'FaceAlpha',alpha)
        material shiny

        %Loop to draw each cylindrical segment for tube
        for i=starttube:endtube-1
            basecirc_trans = reshape(g(i,:),4,4)*basecirc; %current frame circle points
            basecirc_trans_ahead = reshape(g(i+1,:),4,4)*basecirc; %next frame circle points

            %loop to draw each square patch for this segment
            for k=1:radial_pts-1
                xedge = [basecirc_trans(1,k),basecirc_trans(1,k+1),basecirc_trans_ahead(1,k+1),basecirc_trans_ahead(1,k)];
                yedge = [basecirc_trans(2,k),basecirc_trans(2,k+1),basecirc_trans_ahead(2,k+1),basecirc_trans_ahead(2,k)];
                zedge = [basecirc_trans(3,k),basecirc_trans(3,k+1),basecirc_trans_ahead(3,k+1),basecirc_trans_ahead(3,k)];
                patch(xedge,yedge,zedge,color,'EdgeAlpha',0,'FaceAlpha',alpha)
            end
        end

        %circle points at tube end
        basecirc_trans = reshape(g(endtube,:),4,4)*basecirc;
        %draw patches to fill in the circle
        patch(basecirc_trans(1,:),basecirc_trans(2,:),basecirc_trans(3,:),color,'EdgeAlpha',.3,'FaceAlpha',alpha)
        material shiny
    end

    %% Projections
    if options.projections
        plot3(g(:,13), ax.YLim(1)*ones(size(g, 1)), g(:,15), 'LineWidth', 2, 'Color',[0 1 0]); % project in x-z axis
        plot3(ax.XLim(1)*ones(size(g, 1)), g(:,14), g(:,15), 'LineWidth', 2, 'Color',[1 0 0]); % project in y-z axis
        plot3(g(:,13), g(:,14), zeros(size(g, 1)), 'LineWidth', 2, 'Color',[0 0 1]); % project in x-y axis
    end

    %% base plate
    if options.baseplate
        color = [1 1 1]*0.9;
        squaresize = 0.02;
        thickness = 0.001;
        patch([-1 1 1 -1]*squaresize,[-1 -1 1 1]*squaresize,[-1 -1 -1 -1]*thickness,color)
        patch([-1 1 1 -1]*squaresize,[-1 -1 1 1]*squaresize,[0 0 0 0],color)
        patch([1 1 1 1]*squaresize,[-1 -1 1 1]*squaresize,[-1 0 0 -1]*thickness,color)
        patch([-1 -1 -1 -1]*squaresize,[-1 -1 1 1]*squaresize,[-1 0 0 -1]*thickness,color)
        patch([-1 1 1 -1]*squaresize,[-1 -1 -1 -1]*squaresize,[-1 -1 0 0]*thickness,color)
        patch([-1 1 1 -1]*squaresize,[1 1 1 1]*squaresize,[-1 -1 0 0]*thickness,color)
    end

    %% Coordinate Frames
    if options.tipframe && ~options.segframe
        quiver3(g(end,13),g(end,14),g(end,15),g(end,1),g(end,2),g(end,3),0.01,'LineWidth',3,'Color',[1 0 0]);
        quiver3(g(end,13),g(end,14),g(end,15),g(end,5),g(end,6),g(end,7),0.01,'LineWidth',3,'Color',[0 1 0])
        quiver3(g(end,13),g(end,14),g(end,15),g(end,9),g(end,10),g(end,11),0.01,'LineWidth',3,'Color',[0 0 1]);
    end  
    
    if options.segframe
        for i=1:numtubes
            quiver3(g(tube_end(i),13),g(tube_end(i),14),g(tube_end(i),15),g(tube_end(i),1),g(tube_end(i),2),g(tube_end(i),3),0.01,'LineWidth',3,'Color',[1 0 0]);
            quiver3(g(tube_end(i),13),g(tube_end(i),14),g(tube_end(i),15),g(tube_end(i),5),g(tube_end(i),6),g(tube_end(i),7),0.01,'LineWidth',3,'Color',[0 1 0])
            quiver3(g(tube_end(i),13),g(tube_end(i),14),g(tube_end(i),15),g(tube_end(i),9),g(tube_end(i),10),g(tube_end(i),11),0.01,'LineWidth',3,'Color',[0 0 1]);
        end
    end

    if options.baseframe
        quiver3(0,0,0,1,0,0,0.01,'LineWidth',3,'Color',[1 0 0]);
        quiver3(0,0,0,0,1,0,0.01,'LineWidth',3,'Color',[0 1 0]);
        quiver3(0,0,0,0,0,1,0.01,'LineWidth',3,'Color',[0 0 1]);
    end


